benchmarks/README.rst

```diff
@@ -1,16 +1,76 @@
-Performance Testing
-===================
+Benchmarks
+----------
 
-New scenario
-------------
+These benchmarks are intended to provide stable and reproducible measurements of the performance characteristics of the ``ddtrace`` library. A scenario is defined using a simple Python framework. Docker is used to build images for the execution of scenarios against different versions of ``ddtrace``.
 
-Create a new directory for a scenario which includes a ``scenario.py`` script to be run. Add any additional dependencies in a ``requirements_scenario.txt``.
+.. _framework:
 
-First you need to build the scenario image::
+Framework
+^^^^^^^^^
 
-  docker build -t perf-<scenario> --build-arg SCENARIO=<scenario> -f base/Dockerfile .
+A scenario requires:
 
-You can now run the scenario image with two versions of the library. The environment variables ``DDTRACE_INSTALL_{V1,V2}`` can be set to a PEP 508 specification or a git url:
+* ``scenario.py``: implements a class for running a benchmark
+* ``config.yaml``: specifies one or more sets of configuration variables for the benchmark
+* ``requirements_scenario.txt``: any additional dependencies
 
-  docker run -it --rm -e DDTRACE_INSTALL_V1="ddtrace" -e DDTRACE_INSTALL_V2="ddtrace==0.50.0" perf-<scenario>
-  docker run -it --rm -e DDTRACE_INSTALL_V1=git+https://github.com/Datadog/dd-trace-py@master -e DDTRACE_INSTALL_V2=git+https://github.com/Datadog/dd-trace-py@v0.50.0 perf-<scenario>
+The scenario class inherits from ``bm.Scenario`` and includes the configurable variables using ``bm.var``. The execution of the benchmark uses the ``run()`` generator function to yield a function that will handle the execution of a specified number of loops. The scenario class must also be decorated with ``@bm.register`` to ensure that it is run.
+
+Example
+~~~~~~~
+
+``scenario.py``
++++++++++++++++
+
+::
+
+  import bm
+
+
+  @bm.register
+  class MyScenario(bm.Scenario):
```

Comment:
you could use a metaclass to auto-register any class that inherits from `bm.Scenario`

(or.. can a decorator add a metaclass/override inheritance for the class it wraps?)

Reply:
This doesn't look great, but the reason why we have the `register` decorator is to sweep any dependency on `attr` under the rug.

Reply:
yeah, but we can do that via a Metaclass on `bm.Scenario` as well

since we can modify the MRO before actually creating the class.

Reply:
Yep, we'd have to include wrapping the class in `attr.s` if we were to register it in some metaclass-assisted hook.

Reply:
again, reminder... this is a neat alternative approach, not blocking on getting this PR merged.